### ISSUE TITLE
Question on padding_side "left" vs "right"

### ISSUE THREAD

**Issue by hanyin88** [tloen/alpaca-lora#514]
In the finetune.py, we set padding_side="left". I realize this is to enable batched inference and generally speaking I think it make sense for a decoder model. 

However, it seems the default padding_side is "right" when directly loading from "decapoda-research/llama-7b-hf" (plus the truncation_side is also "right"). Also in the original[ alpaca work](https://github.com/tatsu-lab/stanford_alpaca/blob/main/train.py), they set the padding_side to "right" as well. May I ask why there's a difference here?

For my own project, I am working on a Sequence Classification task, where the first few sentences matters the most. I am wondering if it makes more sense to set padding side to "left" in my case to preserve more positional info. 

Also to add that in generate.py, it seems we loaded the default LLaMA Tokenizer from base model which would have padding_side = "right".

Many thanks for the kind help.

@chrisociepa @ElleLeonne hoping to tag your guys given the kind answer in this [post](https://github.com/tloen/alpaca-lora/issues/277)

**Comment by chrisociepa** [comment]
Right padding makes sense when you batching, because you are supposed to add more (generated) text on the right side, not the left side. When finetuning with the attention mask you are masking (prompt) tokens on the left side to focus on expected responses but this is different than the padding and might be be sometimes confusing.

**Comment by hanyin88** [comment]
> Right padding makes sense when you batching, because you are supposed to add more (generated) text on the right side, not the left side. When finetuning with the attention mask you are masking (prompt) tokens on the left side to focus on expected responses but this is different than the padding and might be be sometimes confusing.

Thanks for the rapid reply Chris and this is super enlightening!

Am I understanding you right here? For instruct-based (prompt-based) finetuning, it make sense to pad on the left side (for the clear reason your outlined!). However for general batching purpose or for a different fine-tuning process (for example a text classification task where I only use text as input without prompt), it would make sense to pad on the right side. I should mention in my own project I seems to get better results (lower eval loss) when padding on the right! :)

If so, would you mind commenting on why in the original Alpaca work they padded on the right side (at prompt-based fine tuning stage)? 

Also for future readers, to make things more confusing, in the official tutorial fine-tuning GPT for [sequence classification](https://huggingface.co/docs/peft/task_guides/ptuning-seq-classification), the set padding_side to "left"....

Many thanks! @chrisociepa

**Comment by chrisociepa** [comment]
I might be wrong but I think the LLaMA model and its self-attention assumes the right padding. You can see scaled dot product attention in pytorch [here](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) with [tril](https://pytorch.org/docs/stable/generated/torch.tril.html) to get some sense of it. Assuming this if you left padding, you basically truncate (I simplify here) your prompt and the model can easier learn, so the loss is getting lower faster. Again, I am not 100% sure about this explanation but this is my understanding of your case.

**Comment by sarthak247** [comment]
I am still confused about this and believe this is a major issue which a lot of people face as I have seen similar problems in other repos as well like HF and PyTorch. I really wish there was some easy explanation for this as why some models use Left or Right Padding and why HF explicitly says to use left padding when in actual LLama tokenizer says to use right tokenizer. It can get so confusing at times

**Comment by zhzhangcc** [comment]
This padding side issue gets complicated because of the `position_ids`, which range from 0 -> max_tokens if you don't pass it explicitly, regardless of which side you pad. Meaning if you pad right you can have positions [0, 1, 2, 3, 4] for your tokens, but if you pad left, you can have positions [4090, 4091, 4091, 4093, 4095] instead.

I think this is why performance on Llama models is somehow worse if you pad left instead of right.

**Comment by Boltzmachine** [comment]
Would `position_ids` be a problem if the LLM uses RoPE, where the results only depend on the relative position?
